File: apollo-integration/src/test/java/com/apollographql/apollo/ApolloInterceptorTest.java
package com.apollographql.apollo;


import android.support.annotation.NonNull;

import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.cache.normalized.Record;
import com.apollographql.apollo.exception.ApolloException;
import com.apollographql.apollo.exception.ApolloParseException;
import com.apollographql.apollo.integration.normalizer.EpisodeHeroNameQuery;
import com.apollographql.apollo.integration.normalizer.type.Episode;
import com.apollographql.apollo.interceptor.ApolloInterceptor;
import com.apollographql.apollo.interceptor.ApolloInterceptorChain;
import com.apollographql.apollo.rx2.Rx2Apollo;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import java.io.IOException;
import java.util.Collections;
import java.util.concurrent.Executor;
import java.util.concurrent.TimeoutException;

import javax.annotation.Nonnull;

import io.reactivex.functions.Predicate;
import okhttp3.Dispatcher;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Protocol;
import okhttp3.Request;
import okhttp3.ResponseBody;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;

import static com.apollographql.apollo.Utils.assertResponse;
import static com.apollographql.apollo.Utils.enqueueAndAssertResponse;
import static com.apollographql.apollo.interceptor.ApolloInterceptor.InterceptorResponse;
import static com.google.common.truth.Truth.assertThat;

public class ApolloInterceptorTest {
  private static final String FILE_EPISODE_HERO_NAME_WITH_ID = "EpisodeHeroNameResponseWithId.json";
  private static final String FILE_EPISODE_HERO_NAME_CHANGE = "EpisodeHeroNameResponseNameChange.json";

  private ApolloClient client;
  @Rule public final MockWebServer server = new MockWebServer();
  private OkHttpClient okHttpClient;

  @Before
  public void setup() {
    okHttpClient = new OkHttpClient.Builder()
        .dispatcher(new Dispatcher(Utils.immediateExecutorService()))
        .build();
  }

  @Test
  public void asyncApplicationInterceptorCanShortCircuitResponses() throws Exception {
    server.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID));
    EpisodeHeroNameQuery query = createHeroNameQuery();
    final InterceptorResponse expectedResponse = prepareInterceptorResponse(query);
    ApolloInterceptor interceptor = createShortcutInterceptor(expectedResponse);

    client = createApolloClient(interceptor);
    Rx2Apollo.from(client.query(query)).test()
        .assertValue(expectedResponse.parsedResponse.get());
  }

  @Test
  public void asyncApplicationInterceptorRewritesResponsesFromServer() throws Exception {
    server.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID));
    EpisodeHeroNameQuery query = createHeroNameQuery();
    final InterceptorResponse rewrittenResponse = prepareInterceptorResponse(query);
    ApolloInterceptor interceptor = new ApolloInterceptor() {
      @Override
      public void interceptAsync(@Nonnull InterceptorRequest request, @Nonnull ApolloInterceptorChain chain,
          @Nonnull Executor dispatcher, @Nonnull final CallBack callBack) {
        chain.proceedAsync(request, dispatcher, new CallBack() {
          @Override public void onResponse(@Nonnull InterceptorResponse response) {
            callBack.onResponse(rewrittenResponse);
          }

          @Override public void onFailure(@Nonnull ApolloException e) {
            throw new RuntimeException(e);
          }

          @Override public void onCompleted() {
            callBack.onCompleted();
          }

          @Override public void onFetch(FetchSourceType sourceType) {
            callBack.onFetch(sourceType);
          }
        });
      }

      @Override public void dispose() {

      }
    };

    client = createApolloClient(interceptor);
    Rx2Apollo.from(client.query(query)).test()
        .assertValue(rewrittenResponse.parsedResponse.get());
  }

  @Test
  public void asyncApplicationInterceptorThrowsApolloException() throws Exception {
    final String message = "ApolloException";
    EpisodeHeroNameQuery query = createHeroNameQuery();
    ApolloInterceptor interceptor = new ApolloInterceptor() {
      @Override
      public void interceptAsync(@Nonnull InterceptorRequest request, @Nonnull ApolloInterceptorChain chain,
          @Nonnull Executor dispatcher, @Nonnull CallBack callBack) {
        ApolloException apolloException = new ApolloParseException(message);
        callBack.onFailure(apolloException);
      }

      @Override public void dispose() {

      }
    };

    client = createApolloClient(interceptor);
    Rx2Apollo.from(client.query(query))
        .test()
        .assertError(new Predicate<Throwable>() {
          @Override public boolean test(Throwable throwable) throws Exception {
            return message.equals(throwable.getMessage()) && throwable instanceof ApolloParseException;
          }
        });
  }

  @Test
  public void asyncApplicationInterceptorThrowsRuntimeException() throws TimeoutException, InterruptedException {
    final String message = "RuntimeException";
    EpisodeHeroNameQuery query = createHeroNameQuery();
    ApolloInterceptor interceptor = new ApolloInterceptor() {
      @Override
      public void interceptAsync(@Nonnull InterceptorRequest request, @Nonnull ApolloInterceptorChain chain,
          @Nonnull Executor dispatcher, @Nonnull CallBack callBack) {
        dispatcher.execute(new Runnable() {
          @Override public void run() {
            throw new RuntimeException(message);
          }
        });
      }

      @Override public void dispose() {

      }
    };

    client = createApolloClient(interceptor);

    Rx2Apollo.from(client.query(query)).test()
        .assertError(new Predicate<Throwable>() {
          @Override public boolean test(Throwable throwable) throws Exception {
            return throwable instanceof RuntimeException && message.equals(throwable.getMessage());
          }
        });
  }

  @Test
  public void asyncApplicationInterceptorReturnsNull() throws TimeoutException, InterruptedException {
    EpisodeHeroNameQuery query = createHeroNameQuery();
    ApolloInterceptor interceptor = new ApolloInterceptor() {
      @Override
      public void interceptAsync(@Nonnull InterceptorRequest request, @Nonnull ApolloInterceptorChain chain,
          @Nonnull Executor dispatcher, @Nonnull final CallBack callBack) {
        dispatcher.execute(new Runnable() {
          @Override public void run() {
            callBack.onResponse(null);
          }
        });
      }

      @Override public void dispose() {

      }
    };

    client = createApolloClient(interceptor);

    Rx2Apollo.from(client.query(query)).test()
        .assertError(new Predicate<Throwable>() {
          @Override public boolean test(Throwable throwable) throws Exception {
            return throwable instanceof NullPointerException;
          }
        });
  }

  @Test
  public void applicationInterceptorCanMakeMultipleRequestsToServer() throws Exception {
    server.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_CHANGE));
    EpisodeHeroNameQuery query = createHeroNameQuery();
    ApolloInterceptor interceptor = createChainInterceptor();

    client = createApolloClient(interceptor);

    enqueueAndAssertResponse(
        server,
        FILE_EPISODE_HERO_NAME_WITH_ID,
        client.query(query),
        new Predicate<Response<EpisodeHeroNameQuery.Data>>() {
          @Override public boolean test(Response<EpisodeHeroNameQuery.Data> response) throws Exception {
            assertThat(response.data().hero().name()).isEqualTo("Artoo");
            return true;
          }
        }
    );
  }

  @Test
  public void onShortCircuitingResponseSubsequentInterceptorsAreNotCalled() throws IOException, ApolloException {
    EpisodeHeroNameQuery query = createHeroNameQuery();
    final InterceptorResponse expectedResponse = prepareInterceptorResponse(query);

    ApolloInterceptor firstInterceptor = createShortcutInterceptor(expectedResponse);
    ApolloInterceptor secondInterceptor = createChainInterceptor();

    client = ApolloClient.builder()
        .serverUrl(server.url("/"))
        .okHttpClient(okHttpClient)
        .addApplicationInterceptor(firstInterceptor)
        .addApplicationInterceptor(secondInterceptor)
        .build();

    assertResponse(
        client.query(query),
        new Predicate<Response<EpisodeHeroNameQuery.Data>>() {
          @Override public boolean test(Response<EpisodeHeroNameQuery.Data> response) throws Exception {
            assertThat(expectedResponse.parsedResponse.get()).isEqualTo(response);
            return true;
          }
        }
    );
  }

  @Test
  public void onApolloCallCanceledAsyncApolloInterceptorIsDisposed() throws ApolloException, TimeoutException,
      InterruptedException, IOException {
    server.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID));

    EpisodeHeroNameQuery query = createHeroNameQuery();
    SpyingApolloInterceptor interceptor = new SpyingApolloInterceptor();

    Utils.TestExecutor testExecutor = new Utils.TestExecutor();
    client = createApolloClient(interceptor, testExecutor);

    ApolloCall<EpisodeHeroNameQuery.Data> apolloCall = client.query(query);

    apolloCall.enqueue(new ApolloCall.Callback<EpisodeHeroNameQuery.Data>() {
      @Override public void onResponse(@Nonnull Response<EpisodeHeroNameQuery.Data> response) {
      }

      @Override public void onFailure(@Nonnull ApolloException e) {
      }
    });
    apolloCall.cancel();
    testExecutor.triggerActions();
    assertThat(interceptor.isDisposed).isTrue();
  }

  @NonNull private EpisodeHeroNameQuery createHeroNameQuery() {
    return EpisodeHeroNameQuery
        .builder()
        .episode(Episode.EMPIRE)
        .build();
  }

  private ApolloClient createApolloClient(ApolloInterceptor interceptor, Executor dispatcher) {
    return ApolloClient.builder()
        .serverUrl(server.url("/"))
        .okHttpClient(okHttpClient)
        .addApplicationInterceptor(interceptor)
        .dispatcher(dispatcher)
        .build();
  }

  private ApolloClient createApolloClient(ApolloInterceptor interceptor) {
    return createApolloClient(interceptor, Utils.immediateExecutor());
  }

  @NonNull private InterceptorResponse prepareInterceptorResponse(EpisodeHeroNameQuery query) {
    Request request = new Request.Builder()
        .url(server.url("/"))
        .build();

    okhttp3.Response okHttpResponse = new okhttp3.Response.Builder()
        .request(request)
        .protocol(Protocol.HTTP_2)
        .code(200)
        .message("Intercepted")
        .body(ResponseBody.create(MediaType.parse("text/plain; charset=utf-8"), "fakeResponse"))
        .build();

    Response<EpisodeHeroNameQuery.Data> apolloResponse = Response.<EpisodeHeroNameQuery.Data>builder(query).build();

    return new InterceptorResponse(okHttpResponse,
        apolloResponse, Collections.<Record>emptyList());
  }

  private MockResponse mockResponse(String fileName) throws IOException {
    return new MockResponse().setChunkedBody(Utils.readFileToString(getClass(), "/" + fileName), 32);
  }

  private static class SpyingApolloInterceptor implements ApolloInterceptor {

    volatile boolean isDisposed = false;

    @Override
    public void interceptAsync(@Nonnull InterceptorRequest request, @Nonnull ApolloInterceptorChain chain, @Nonnull Executor dispatcher, @Nonnull CallBack callBack) {
      chain.proceedAsync(request, dispatcher, callBack);
    }

    @Override public void dispose() {
      isDisposed = true;
    }
  }

  @NonNull private static ApolloInterceptor createChainInterceptor() {
    return new ApolloInterceptor() {
      @Override
      public void interceptAsync(@Nonnull InterceptorRequest request, @Nonnull ApolloInterceptorChain chain,
          @Nonnull Executor dispatcher, @Nonnull CallBack callBack) {
        chain.proceedAsync(request, dispatcher, callBack);
      }

      @Override public void dispose() {

      }
    };
  }

  @NonNull private static ApolloInterceptor createShortcutInterceptor(final InterceptorResponse expectedResponse) {
    return new ApolloInterceptor() {
      @Override
      public void interceptAsync(@Nonnull InterceptorRequest request, @Nonnull ApolloInterceptorChain chain,
          @Nonnull Executor dispatcher, @Nonnull CallBack callBack) {
        callBack.onResponse(expectedResponse);
      }

      @Override public void dispose() {
      }
    };
  }
}

File: apollo-integration/src/test/kotlin/com/apollographql/apollo3/ApolloInterceptorTest.kt
package com.apollographql.apollo3

import com.apollographql.apollo3.Utils.assertResponse
import com.apollographql.apollo3.Utils.enqueueAndAssertResponse
import com.apollographql.apollo3.Utils.immediateExecutor
import com.apollographql.apollo3.Utils.immediateExecutorService
import com.apollographql.apollo3.Utils.readFileToString
import com.apollographql.apollo3.api.Input
import com.apollographql.apollo3.api.Response
import com.apollographql.apollo3.api.Response.Companion.builder
import com.apollographql.apollo3.exception.ApolloException
import com.apollographql.apollo3.exception.ApolloParseException
import com.apollographql.apollo3.integration.normalizer.EpisodeHeroNameQuery
import com.apollographql.apollo3.integration.normalizer.type.Episode
import com.apollographql.apollo3.interceptor.ApolloInterceptor
import com.apollographql.apollo3.interceptor.ApolloInterceptor.*
import com.apollographql.apollo3.interceptor.ApolloInterceptorChain
import com.apollographql.apollo3.rx2.Rx2Apollo
import com.google.common.truth.Truth
import com.google.common.truth.Truth.assertThat
import io.reactivex.functions.Predicate
import okhttp3.*
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.Before
import org.junit.Test
import java.io.IOException
import java.util.concurrent.Executor
import java.util.concurrent.TimeoutException

class ApolloInterceptorTest {
  private lateinit var client: ApolloClient

  val server = MockWebServer()

  private var okHttpClient: OkHttpClient? = null

  @Before
  fun setup() {
    okHttpClient = OkHttpClient.Builder()
        .dispatcher(Dispatcher(immediateExecutorService()))
        .build()
  }

  @Test
  @Throws(Exception::class)
  fun asyncApplicationInterceptorCanShortCircuitResponses() {
    server.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID))
    val query = createHeroNameQuery()
    val expectedResponse = prepareInterceptorResponse(query)
    val interceptor = createShortcutInterceptor(expectedResponse)
    client = createApolloClient(interceptor)
    Rx2Apollo.from(client.query(query)).test().assertValue(expectedResponse.parsedResponse.get() as Response<EpisodeHeroNameQuery.Data>)
  }

  @Test
  @Throws(Exception::class)
  fun asyncApplicationInterceptorRewritesResponsesFromServer() {
    server.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID))
    val query = createHeroNameQuery()
    val rewrittenResponse = prepareInterceptorResponse(query)
    val interceptor: ApolloInterceptor = object : ApolloInterceptor {
      override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain,
                                  dispatcher: Executor, callBack: CallBack) {
        chain.proceedAsync(request, dispatcher, object : CallBack {
          override fun onResponse(response: InterceptorResponse) {
            callBack.onResponse(rewrittenResponse)
          }

          override fun onFailure(e: ApolloException) {
            throw RuntimeException(e)
          }

          override fun onCompleted() {
            callBack.onCompleted()
          }

          override fun onFetch(sourceType: FetchSourceType) {
            callBack.onFetch(sourceType)
          }
        })
      }

      override fun dispose() {}
    }
    client = createApolloClient(interceptor)
    Rx2Apollo.from(client.query(query)).test().assertValue(rewrittenResponse.parsedResponse.get() as Response<EpisodeHeroNameQuery.Data>)
  }

  @Test
  @Throws(Exception::class)
  fun asyncApplicationInterceptorThrowsApolloException() {
    val message = "ApolloException"
    val query = createHeroNameQuery()
    val interceptor: ApolloInterceptor = object : ApolloInterceptor {
      override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain,
                                  dispatcher: Executor, callBack: CallBack) {
        val apolloException: ApolloException = ApolloParseException(message)
        callBack.onFailure(apolloException)
      }

      override fun dispose() {}
    }
    client = createApolloClient(interceptor)
    Rx2Apollo.from(client.query(query))
        .test()
        .assertError { throwable -> message == throwable.message && throwable is ApolloParseException }
  }

  @Test
  @Throws(TimeoutException::class, InterruptedException::class)
  fun asyncApplicationInterceptorThrowsRuntimeException() {
    val message = "RuntimeException"
    val query = createHeroNameQuery()
    val interceptor: ApolloInterceptor = object : ApolloInterceptor {
      override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain,
                                  dispatcher: Executor, callBack: CallBack) {
        dispatcher.execute { throw RuntimeException(message) }
      }

      override fun dispose() {}
    }
    client = createApolloClient(interceptor)
    Rx2Apollo.from(client.query(query)).test().assertError { throwable -> throwable is RuntimeException && message == throwable.message }
  }

  @Test
  @Throws(Exception::class)
  fun applicationInterceptorCanMakeMultipleRequestsToServer() {
    server.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_CHANGE))
    val query = createHeroNameQuery()
    val interceptor = createChainInterceptor()
    client = createApolloClient(interceptor)
    enqueueAndAssertResponse(
        server,
        FILE_EPISODE_HERO_NAME_WITH_ID,
        client.query(query)
    ) { (_, data) ->
      assertThat(data!!.hero?.name).isEqualTo("Artoo")
      true
    }
  }

  @Test
  @Throws(IOException::class, ApolloException::class)
  fun onShortCircuitingResponseSubsequentInterceptorsAreNotCalled() {
    val query = createHeroNameQuery()
    val expectedResponse = prepareInterceptorResponse(query)
    val firstInterceptor = createShortcutInterceptor(expectedResponse)
    val secondInterceptor = createChainInterceptor()
    client = ApolloClient.builder()
        .serverUrl(server.url("/"))
        .okHttpClient(okHttpClient!!)
        .addApplicationInterceptor(firstInterceptor)
        .addApplicationInterceptor(secondInterceptor)
        .build()
    assertResponse(
        client.query(query)
    ) { response ->
      assertThat(expectedResponse.parsedResponse.get()).isEqualTo(response)
    }
  }

  @Test
  @Throws(ApolloException::class, TimeoutException::class, InterruptedException::class, IOException::class)
  fun onApolloCallCanceledAsyncApolloInterceptorIsDisposed() {
    server.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID))
    val query = createHeroNameQuery()
    val interceptor = SpyingApolloInterceptor()
    val testExecutor = Utils.TestExecutor()
    client = createApolloClient(interceptor, testExecutor)
    val apolloCall: ApolloCall<EpisodeHeroNameQuery.Data> = client.query(query)
    apolloCall.enqueue(object : ApolloCall.Callback<EpisodeHeroNameQuery.Data>() {
      override fun onResponse(response: Response<EpisodeHeroNameQuery.Data>) {}
      override fun onFailure(e: ApolloException) {}
    })
    apolloCall.cancel()
    testExecutor.triggerActions()
    Truth.assertThat(interceptor.isDisposed).isTrue()
  }

  private fun createHeroNameQuery(): EpisodeHeroNameQuery {
    return EpisodeHeroNameQuery(episode = Input.present(Episode.EMPIRE))
  }

  private fun createApolloClient(interceptor: ApolloInterceptor, dispatcher: Executor = immediateExecutor()): ApolloClient {
    return ApolloClient.builder()
        .serverUrl(server.url("/"))
        .okHttpClient(okHttpClient!!)
        .addApplicationInterceptor(interceptor)
        .dispatcher(dispatcher)
        .build()
  }

  private fun prepareInterceptorResponse(query: EpisodeHeroNameQuery): InterceptorResponse {
    val request = Request.Builder()
        .url(server.url("/"))
        .build()
    val okHttpResponse = okhttp3.Response.Builder()
        .request(request)
        .protocol(Protocol.HTTP_2)
        .code(200)
        .message("Intercepted")
        .body(ResponseBody.create(MediaType.parse("text/plain; charset=utf-8"), "fakeResponse"))
        .build()
    val apolloResponse = builder<EpisodeHeroNameQuery.Data>(query).build()
    return InterceptorResponse(okHttpResponse, apolloResponse)
  }

  @Throws(IOException::class)
  private fun mockResponse(fileName: String): MockResponse {
    return MockResponse().setChunkedBody(readFileToString(javaClass, "/$fileName"), 32)
  }

  private class SpyingApolloInterceptor : ApolloInterceptor {
    @Volatile
    var isDisposed = false
    override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain, dispatcher: Executor, callBack: CallBack) {
      chain.proceedAsync(request, dispatcher, callBack)
    }

    override fun dispose() {
      isDisposed = true
    }
  }

  companion object {
    private const val FILE_EPISODE_HERO_NAME_WITH_ID = "EpisodeHeroNameResponseWithId.json"
    private const val FILE_EPISODE_HERO_NAME_CHANGE = "EpisodeHeroNameResponseNameChange.json"
    private fun createChainInterceptor(): ApolloInterceptor {
      return object : ApolloInterceptor {
        override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain,
                                    dispatcher: Executor, callBack: CallBack) {
          chain.proceedAsync(request, dispatcher, callBack)
        }

        override fun dispose() {}
      }
    }

    private fun createShortcutInterceptor(expectedResponse: InterceptorResponse): ApolloInterceptor {
      return object : ApolloInterceptor {
        override fun interceptAsync(request: InterceptorRequest, chain: ApolloInterceptorChain,
                                    dispatcher: Executor, callBack: CallBack) {
          callBack.onResponse(expectedResponse)
        }

        override fun dispose() {}
      }
    }
  }
}
